Mh 254 cancel cf document run

sharrisoncaktus opened the pull request:
This PR adds the ability to Abort and Update CloudFactoryDocumentRuns. Highlights:

* new celery tasks: CloudFactoryAbortDocumentRun and CloudFactoryUpdateDocumentRun. These tasks only send to the CF API when their status indicates that something should be done.
* the admin for document runs (list view) now has actions for updating and aborting runs. it also links back to the associated AD (both list view and individual view)
* deleting a DocumentRun automatically triggers an abort task for any run object associated with the DocumentURL (and these tasks only send to the API when appropriate).
* CloudFactoryDocumentRun.STATUS_CHOICES adds a new STATUS_ERROR for any situation when the response.content is not json-parsable and is expected to be.
* normalization of all the STATUS_CHOICES and explanation of what each status choice means in the CloudFactoryDocumentRun docstring.
* _darn_ good test coverage of these CF API tasks IIDSSM (if I dare say so myself)

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-254--cancel-cf-ad-run

* myhpom-develop:
  Puts document urls into an AD admin area (MH-278)

sharrisoncaktus committed:
[mh-254] when a document has been deleted before the run has been posted to CF, set the status to 'DELETED' rather than 'ABORTED' so as to avoid confusion with the 'Aborted' status of units in the run.

sharrisoncaktus committed:
[mh-254] CloudFactoryRun and CloudFactoryUnit need methods to save the response data from CF. the Run version of this needs to be wrapped in a transaction.

sharrisoncaktus committed:
[mh-254] new tasks to Cancel and Update a CloudFactoryRun via the CF API

sharrisoncaktus committed:
[mh-254] When a DocumentUrl object is deleted, automatically abort the associated CloudFactoryRuns

sharrisoncaktus committed:
[mh-254] (with previous)

sharrisoncaktus committed:
[mh-254] admin interface for CloudFactoryRuns. Include links to cf_runs from the DocumentUrl admin.

sharrisoncaktus committed:
[mh-254] new tasks and admin actions: Abort Run, Update Run

sharrisoncaktus committed:
Merge branch 'mh-252--celery-setup' into mh-254-cancel-cf-document-run

* mh-252--celery-setup:
  [mh-252] new status choices require updating the migration.
  [mh-252] added 'Aborted' status for the next ticket

sharrisoncaktus committed:
Merge branch 'mh-252--celery-setup' into mh-254-cancel-cf-document-run

* mh-252--celery-setup:
  [mh-252] shoot the strays.

sharrisoncaktus committed:
Merge commit '946ff6462e1cf672543e8b1b6749c039d48409a3' into mh-254-cancel-cf-document-run

* commit '946ff6462e1cf672543e8b1b6749c039d48409a3':

sharrisoncaktus committed:
[mh-254] completed logic for AbortRun and UpdateRun tasks

sharrisoncaktus committed:
[mh-254] documentation for the CloudFactoryDocumentRun.STATUS_CHOICES

sharrisoncaktus committed:
[mh-254] let django parse the .created_at and .processed_at fields automatically. They've not been parsed yet in the test, so the test should just compare strings.

sharrisoncaktus committed:
[mh-254] added run.updated_at to reflect when the run object record was last updated.

sharrisoncaktus committed:
[mh-254] update the run before processing the abort.

sharrisoncaktus committed:
[mh-254] new run status = STATUS_ERROR, used when the response is non-json. This status is now set in the run.save_response_content() method to avoid repeating it everywhere in tasks.

sharrisoncaktus committed:
[mh-254] don't call UpdateRun from within AbortRun, because it's redundant – we can do everything in one API call.

sharrisoncaktus committed:
[mh-254] during AbortRun, save_response_content() for both 202 and 405 statuses, and save the raw response_content on 404.

sharrisoncaktus committed:
[mh-254] All the tests for CloudFactory UpdateRun and AbortRun, and correction to SubmitRun non-json test to reflect new STATUS_ERROR

sharrisoncaktus committed:
[mh-254] add AdvanceDirective link to the CloudFactoryDocumentRun admin page so we can link back

sharrisoncaktus committed:
[mh-254] ensure that any CloudFactoryDocumentRuns associated with a DocumentUrl are aborted when the DocumentUrl is deleted (delete cascades from AdvanceDirective)

sharrisoncaktus committed:
[mh-254] add test for abort DocumentRun when DocumentUrl is deleted.

sharrisoncaktus committed:
[mh-254] flake8 found a test bug!  молодец (good boy), flake8

dsummersl commented on myhpom/tasks.py:
```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
+            try:
+                run_url = "%s/runs/%s" % (settings.CLOUDFACTORY_API_URL, cf_run.run_id)
+                response = requests.get(run_url)
+            except requests.exceptions.ConnectTimeout:
```
There are a lot of request exception possibilities, not just timeout - catch RequestException (see http://docs.python-requests.org/en/master/api/?highlight=requestexception#requests.RequestException) -- which probably means STATUS_TIMEOUT should be something more generic.

sharrisoncaktus replied:
Good idea.

sharrisoncaktus replied:
Good idea.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
+            else:
+                # don't change the run status here -- we haven't learned anything about it,
+                # and maybe there's a response status that we don't know about
+                raise ValueError(
```
We don't know what it is, but its definitely not expected. I'd suggest putting this in some not-okay terminal state.

sharrisoncaktus replied:
Okay, we could use STATUS_ERROR for this.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
+            else:
+                # don't change the run status here -- we haven't learned anything about it,
+                # and maybe there's a response status that we don't know about.
+                raise ValueError(
```
same comment as above...

sharrisoncaktus replied:
STATUS_ERROR

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -94,10 +97,268 @@ def test_connection_timeout(self, reqmock):
+            cf_run.refresh_from_db()
+            self.assertTrue(reqmock.called)
+            self.assertEqual(cf_run.status, initial_status)
+            cf_run.delete()
```
don't need to `delete()` here as the TestCase will clean up created objects. (this applies to several other deletes in other test cases here as well.

sharrisoncaktus replied:
These runs are in a loop inside the test case method, so the TestCase won't clean them up for me, and I'll get errors for re-using the same run_id:
```
root@19d73ebf1f61:/hydroshare# python manage.py test myhpom.tests.test_cloudfactory_tasks
Creating test database for alias 'default'...
EEEEEE........EEEE
======================================================================
ERROR: test_202_invalid (myhpom.tests.test_cloudfactory_tasks.CloudFactoryAbortDocumentRunTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
...
DETAIL:  Key (run_id)=(SOME_RUN) already exists.
```

sharrisoncaktus replied:
It's needed since we're in a  loop.

dsummersl replied:
oh...b/c the run_id is the same. okay, disregard.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -94,10 +97,268 @@ def test_connection_timeout(self, reqmock):
+                text='{}',
+            )
+            self.task(cf_run.id)
+            self.assertTrue(reqmock.called)
```
The `reqmock.called` here doesn't assert anything useful the second time thru the loop. I guess you could user `call_count`? A true mock would let you reset the mock...but I don't see anything in the requests-mock library that lets you do that :/

sharrisoncaktus replied:
hmmm, good point.

The solution in all of these looping test cases is to just instantiate the reqmock object within the loop, so I've updated the tests accordingly.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -66,17 +69,103 @@ def run(self, document_url_id, document_host=None):
+            because it means we have a bug / false assumption somewhere in our system.
+        """
+        cf_run = CloudFactoryDocumentRun.objects.get(id=cf_run_id)
+        if cf_run.status in [
```
One of the primary goals of CloudFactoryDocumentRun is that it tracks the communication between CF and us. By updating the previous record on abort, we’re losing some of that important information. (the ‘update’ task above isn’t in the AC and is more of a developer only tool I’m thinking so maybe its okay that it does an update?)

With that in mind I’d suggest:
 * Adding a field to CloudFactoryDocumentRun like ‘action’: create, and abort.
 * Creating a new CloudFactoryDocumentRun here.

This if block below also brings to mind invarients that we should check for:
 * we wouldn’t expect you to abort a previous run that didn’t finish successfully (this check you’re doing here - this task should fail, and we should get an email.) As in the initial task there should be an initial status set on the run instance that reflects this state.
 * currently, we expect a successful CF process to have no more than one ‘create’ style run and no more than one ‘abort’ style run (until/unless we decide we need to put in some robust retry logic on network failures).

sharrisoncaktus replied:
AC: "Admin user can verify that a document review was cancelled." The abort task doesn't do that verification – it only stores that CF accepted the abort request. But I am envisioning a case where they accept the abort request but are actually already processing the run and finish processing it. So the result is that the processing was not in fact aborted. The "Aborted" status is non-final – it leaves room for a final verification that the abort was finished and the status="Processed". — Maybe the reality with CF is if they accept the abort request, then we can call the run aborted and be done with it.

When we update the response_content field on 202 or 405, we're not losing anything – the .save_response_content() only updates fields that are given. In fact, that method doesn't do anything in this case except save the status, so I've adjusted the task just to do that. This also removes the possibility of an error on non-json response body, since we don't do anything with the body.

So I don't think we need a new run for the "abort". Since the run model tracks the status of one CF run / process, it makes sense to me to keep it as a single object. In the case of errors, it's always possible to re-submit the document in a different run, but that too is a single run object.

The if-block serves the admin users and the rest of us: If an admin selects all the runs and then hits the 'Abort' admin action, it only actually processes those that ought to be processed – namely, the STATUS_PROCESSING and STATUS_REQ_ERROR (was STATUS_TIMEOUT). I include the second of these because we don't know if the run actually "took" at CF during that request — it's potentially a live run, so this gives us a chance to abort any "potential run". 

We could move the if-block to the admin if that makes more sense to you. I just didn't think the admin action should have to know much about the different STATUS_CHOICES in the model.

dsummersl replied:
Hah, thats what I get for not paying attention to the API docs - I didn't realize the 'abort task' was asynchronous and requires a status check. That makes perfect sense in terms of verification then.

We're doing more fire/forget style generally since I don't see a task to check the status after scheduling an abort...which...I think is okay? (rather than scheduling a followup with the CloudFactoryUpdateDocumentRun task to verify that the abort happened)

sharrisoncaktus replied:
Yes, I think it's okay – we can treat STATUS_ABORTED as a final status, but if the admin wants to get an update from CF, they can and it becomes STATUS_PROCESSED with 'Aborted' in the response_content.

dsummersl reviewed (changes requested):
Overall looks good.

I think the abort action needs to create a new run (which I don't foresee being a huge change?).
The test cases are deleting when they don't need to, and they aren't verifying that the request-mocks are called usefully inside the tests that loop.

sharrisoncaktus committed:
[mh-254] PR review #1

dsummersl commented on myhpom/models/document.py:
```diff
@@ -149,3 +150,14 @@ def authorized_client_ip(self, ip_address):
+
+def abort_document_runs_on_delete(sender, instance, using, **kwargs):
+    """Any document runs associated with a DocumentUrl should be aborted at CF on delete."""
+    if instance.cloudfactorydocumentrun_set.exists:
```
This should be `instance.cloudfactorydocumentrun_set.exists()`? Add a test to the model tests that verifies that either that `tasks.CloudFactoryAbourtDocumentRun.delay` is called (mock-like), or that it happened (requests-mock like).

sharrisoncaktus replied:
re/exists: oops, yeah.

sharrisoncaktus replied:
The model tests don't call the tasks at all, so I'm not sure what you're looking for.

sharrisoncaktus replied:
Oh, you mean for the DocumentUrl model, not for the CFDocumentRun model. Ok, that makes sense.

sharrisoncaktus committed:
[mh-254] bugfix

sharrisoncaktus committed:
[mh-254] test that the run-abort task is called when DocumentUrl is deleted

sharrisoncaktus committed:
[mh-254] make sure that the correct task is the one being called (even though it's the only one in this context.)

dsummersl commented on myhpom/tasks.py:
```diff
@@ -57,6 +57,12 @@ def run(self, document_url_id, document_host=None):
+            cf_run.save()
 
         # if the run could not be created (status_code != 201), send support email
         # -- we need to understand why the the run could not be created at CloudFactory.
```
now that I've got all the way out of 'review' mode and into try mode the first thing I stumble on is not setting up my credentials..this block here should save result to response_content here, even though it wasn't the expected response.

dsummersl replied:
Another aspect of this - we should have the actual secrets stored in our secrets files (CLOUDFACTORY_LINE_ID, etc al). They aren't currently.

sharrisoncaktus replied:
The response_content is now saved no matter what during Submit..Run (if there is a response – i.e., not if there is a RequestError, but otherwise).

sharrisoncaktus replied:
I've added the CLOUDFACTORY_* variables to the ansible-vault templates for staging, production, mmhstaging, and mmhproduction.

sharrisoncaktus committed:
[mh=254]

save response.content to run during Submit...Run task. Rename the model .save_response_content() to .save_response_data() for clarity that it goes to 11.

sharrisoncaktus committed:
[mh-254]

add CLOUDFACTORY_* to ansible-vault templates
